### PR TITLE
Fixing region fail over issue for direct tcp calls

### PR DESCRIPTION
**Pull request by simplynaveen20**
Currently for tcp network exception we are doing region failover, which is causing problem for some users where due to some intermittent network issue on BE nodes sdk is sending traffic for next 5 min to other available region.

This PR will limit the region failover only in case of http network failure.

### CODE REVIEW

File: gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java [hunk 1]
```diff
@@ -126,8 +128,13 @@ public ClientRetryPolicy(GlobalEndpointManager globalEndpointManager,
         // Received Connection error (HttpRequestException), initiate the endpoint rediscovery
         if (WebExceptionUtility.isNetworkFailure(e)) {
             if (this.isReadRequest || WebExceptionUtility.isWebExceptionRetriable(e)) {
-                logger.warn("Endpoint not reachable. Will refresh cache and retry. ", e);
-                return this.shouldRetryOnEndpointFailureAsync(this.isReadRequest, false);
+                if (clientException != null && Exceptions.isSubStatusCode(clientException, HttpConstants.SubStatusCodes.GATEWAY_ENDPOINT_UNAVAILABLE)) {
```
Comment (simplynaveen20) [inline gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java:131]: Will do similar check for write in next iteration

File: gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java [hunk 2]
```diff
@@ -63,6 +64,7 @@
     private RetryContext retryContext;
     private ClientSideRequestStatistics clientSideRequestStatistics;
     private AtomicInteger cnt = new AtomicInteger(0);
+    private int serviceUnavailableRetryCount;
```
Comment (kushagraThapar) [inline gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java:67]: Shouldn't this be volatile ?
Please check other retry policies code to make sure we are covering the race conditions.

File: gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java [hunk 3]
```diff
@@ -207,6 +214,36 @@ private ShouldRetryResult shouldRetryOnSessionNotAvailable() {
         return refreshCompletion.andThen(Single.just(ShouldRetryResult.noRetry()));
     }
 
+    private Single<ShouldRetryResult> shouldRetryOnBackendServiceUnavailableAsync(boolean isReadRequest, boolean isWebExceptionRetriable) {
```
Comment (kushagraThapar) [inline gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java:217]: So this method will only be called when the status code is not `10001` ?
which would mean for all other sub status codes ?

Comment (simplynaveen20) [reply gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java:217]: 10001 is now new substatus code which will be there in all gateway request if they face  network error. For others we dont want to do failover

File: gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java (hunk 2 above)
Comment (simplynaveen20) [reply gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java:67]: this should be fine as we already have failoverRetryCount in this class as non volatile

File: direct-impl/src/main/java/com/microsoft/azure/cosmosdb/internal/directconnectivity/GatewayAddressCache.java [hunk 4]
```diff
@@ -333,14 +333,37 @@ private URL getServiceEndpoint() {
...
+                logger.error("Unexpected failure {}", throwable.getMessage(), throwable);
+                return Single.error(throwable);
+            }
+
+            Exception exception = (Exception) throwable;
+            DocumentClientException dce;
+            if (!(exception instanceof DocumentClientException)) {
+                // wrap in DocumentClientException
+                logger.error("Network failure", exception);
+                dce = new DocumentClientException(0, exception);
```
[the remark this answers is not in the document]
Comment (simplynaveen20) [reply direct-impl/src/main/java/com/microsoft/azure/cosmosdb/internal/directconnectivity/GatewayAddressCache.java:355]: We don't have any constructor for exception alone, and this code is carried over from RxGatewayStoreModel which is main class for gateway calls. Ideally address refresh should go through same stack , but it is not in scope of this PR.

File: gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java (hunk 2 above)
Comment (kushagraThapar) [reply gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java:67]: Could that be a bug too, we should verify these race conditions going forward. 
@moderakh - what are your thoughts on `volatile` / `AtomicInteger` / `non-volatile` counter.

File: direct-impl/src/main/java/com/microsoft/azure/cosmosdb/internal/directconnectivity/GatewayAddressCache.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kirankumarkolli) [reply direct-impl/src/main/java/com/microsoft/azure/cosmosdb/internal/directconnectivity/GatewayAddressCache.java:355]: "0" status code will byte back. Lets revisit it.

File: sdk/src/test/java/com/microsoft/azure/cosmosdb/rx/CollectionCrudTest.java [hunk 5]
```diff
@@ -255,7 +255,7 @@ public void replaceCollection(String collectionName, boolean isNameBased)  {
 
         // validate
         ResourceResponseValidator<DocumentCollection> validator = new ResourceResponseValidator.Builder<DocumentCollection>()
-                .indexingMode(IndexingMode.Lazy).build();
+                .indexingMode(IndexingMode.Consistent).build();
```
Comment (kirankumarkolli) [inline sdk/src/test/java/com/microsoft/azure/cosmosdb/rx/CollectionCrudTest.java:258]: Best if it can be a separate PR

Comment (simplynaveen20) [reply sdk/src/test/java/com/microsoft/azure/cosmosdb/rx/CollectionCrudTest.java:258]: Will revert this in next iteration, below is the PR
https://github.com/Azure/azure-cosmosdb-java/pull/354

**Comment by kushagraThapar** [comment]
Discussed the above comments with @simplynaveen20 offline. Good to go from my side.

File: commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java [hunk 6]
```diff
@@ -51,8 +51,8 @@
     private static final long serialVersionUID = 1L;
 
     private final Map<String, String> requestHeaders;
-    private final Map<String, String> responseHeaders;
     private final int statusCode;
+    private volatile Map<String, String> responseHeaders;
```
[the remark this answers is not in the document]
Comment (simplynaveen20) [reply commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java:55]: We have to remove final to set  subStatuscode , and constructor are getting used at some place with null response header.

File: commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java [hunk 7]
```diff
@@ -321,6 +321,14 @@ public String toString() {
                 '}';
     }
 
+	void setSubStatusCode(int subStatusCode) {
+		if (this.responseHeaders == null) {
```
[the remark this answers is not in the document]
Comment (simplynaveen20) [reply commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java:325]: constructor are getting used at some place with null response header.

[the remark this answers is not in the document]
Comment (moderakh) [reply commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java:325]: all the constructors at the end of the day will invoke the following one:

```
    private DocumentClientException(
        int statusCode, String message, Map<String, String> responseHeaders, Throwable cause) {
        super(message, cause);
        this.requestHeaders = new ConcurrentHashMap<>();
        this.responseHeaders = responseHeaders == null
            ? new ConcurrentHashMap<>()
            : new ConcurrentHashMap<>(responseHeaders);
        this.statusCode = statusCode;
    }
```

File: commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (moderakh) [reply commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java:55]: all the constructors are invoking this one which initializes the headers to a non null value. we should stick to the existing constructor and not re-initialize responseHeaders outside of the constructor

https://github.com/Azure/azure-cosmosdb-java/blob/master/commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java#L141-L149

File: commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (simplynaveen20) [reply commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java]: done

File: commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (simplynaveen20) [reply commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java]: done

File: gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java [hunk 8]
```diff
@@ -207,6 +214,36 @@ private ShouldRetryResult shouldRetryOnSessionNotAvailable() {
...
+        if (!isReadRequest && !isWebExceptionRetriable) {
+            logger.warn("shouldRetryOnBackendServiceUnavailableAsync() Not retrying on write with non retriable exception. Retry count = {}", this.serviceUnavailableRetryCount);
+            return Single.just(ShouldRetryResult.noRetry());
+        }
+
+        if (this.serviceUnavailableRetryCount++ > MaxServiceUnavailableRetryCount) {
+			logger.warn("shouldRetryOnBackendServiceUnavailableAsync() Not retrying. Retry count = {}", this.serviceUnavailableRetryCount);
```
[the remark this answers is not in the document]
Comment (simplynaveen20) [reply gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java]: done

File: commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java [hunk 9]
```diff
@@ -51,8 +51,8 @@
     private static final long serialVersionUID = 1L;
 
     private final Map<String, String> requestHeaders;
-    private final Map<String, String> responseHeaders;
     private final int statusCode;
+    private Map<String, String> responseHeaders;
```
Comment (moderakh) [inline commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java:55]: please change to final.

File: gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java [hunk 10]
```diff
@@ -207,6 +214,36 @@ private ShouldRetryResult shouldRetryOnSessionNotAvailable() {
...
+            logger.warn("shouldRetryOnBackendServiceUnavailableAsync() Not retrying on write with non retriable exception. Retry count = {}", this.serviceUnavailableRetryCount);
+            return Single.just(ShouldRetryResult.noRetry());
+        }
+
+        if (this.serviceUnavailableRetryCount++ > MaxServiceUnavailableRetryCount) {
+            logger.warn("shouldRetryOnBackendServiceUnavailableAsync() Not retrying. Retry count = {}", this.serviceUnavailableRetryCount);
+            return Single.just(ShouldRetryResult.noRetry());
+        }
+
+		if (!this.canUseMultipleWriteLocations && ! isReadRequest) {
```
Comment (moderakh) [inline gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java:228]: invalid indentation.

File: gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java [hunk 11]
```diff
@@ -207,6 +214,36 @@ private ShouldRetryResult shouldRetryOnSessionNotAvailable() {
...
+			// Write requests on single master cannot be retried, no other regions available
+			return Single.just(ShouldRetryResult.noRetry());
+		}
+
+		int availablePreferredLocations = this.globalEndpointManager.getPreferredLocationCount();
+		if (availablePreferredLocations <= 1) {
+			logger.warn("shouldRetryOnServiceUnavailable() Not retrying. No other regions available for the request. AvailablePreferredLocations = {}", availablePreferredLocations);
```
Comment (moderakh) [inline gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java:235]: invalid indentation

File: gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java (hunk 2 above)
Comment (moderakh) [inline gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java:67]: serviceUnavailableRetryCount is not set by anything in your code. 

Is some logic missing? or do we need this at all?

Comment (moderakh) [reply gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java:67]: serviceUnavailableRetryCount is not set by anything in your code.

Is some logic missing? or do we need this at all?

File: gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java [hunk 12]
```diff
@@ -207,6 +214,36 @@ private ShouldRetryResult shouldRetryOnSessionNotAvailable() {
         return refreshCompletion.andThen(Single.just(ShouldRetryResult.noRetry()));
     }
 
+    private Single<ShouldRetryResult> shouldRetryOnBackendServiceUnavailableAsync(boolean isReadRequest, boolean isWebExceptionRetriable) {
+        if (!isReadRequest && !isWebExceptionRetriable) {
+            logger.warn("shouldRetryOnBackendServiceUnavailableAsync() Not retrying on write with non retriable exception. Retry count = {}", this.serviceUnavailableRetryCount);
+            return Single.just(ShouldRetryResult.noRetry());
+        }
+
+        if (this.serviceUnavailableRetryCount++ > MaxServiceUnavailableRetryCount) {
```
Comment (moderakh) [inline gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java:223]: serviceUnavailableRetryCount value always be zero. you are not setting/incrementing it anywhere.

File: gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java (hunk 2 above)
Comment (moderakh) [reply gateway/src/main/java/com/microsoft/azure/cosmosdb/rx/internal/ClientRetryPolicy.java:67]: volatile is fine. as the retry although happening on different threads will not happen concurrently.

File: commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java (hunk 9 above)
Comment (simplynaveen20) [reply commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java:55]: because of set sub status method, it cant be final

**Comment by simplynaveen20** [comment]
> looks good to me thanks Naveen.
> 
> Once scenario we haven't covered is when address resolution request to GW fails.
> This will cause a region failover, but instead we should be more lenient toward GW address all and not fail over the region.

Yes as discussed , its not only address resolution, we need to think and change region fail over for all GW calls.

File: commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java (hunk 9 above)
Comment (moderakh) [reply commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java:55]: you are not initializing responseHeaders in setSubstatus method.

So it should be possible to make this final.

Comment (simplynaveen20) [reply commons/src/main/java/com/microsoft/azure/cosmosdb/DocumentClientException.java]: sorry my mistake i thought it otherway , thanks

File: commons/src/main/java/com/microsoft/azure/cosmosdb/internal/HttpConstants.java [hunk 13]
```diff
@@ -323,6 +323,9 @@
 
         // 404: LSN in session token is higher
         public static final int READ_SESSION_NOT_AVAILABLE = 1002;
+
+        // Client generated gateway network error substatus
+        public static final int GATEWAY_ENDPOINT_UNAVAILABLE = 10001;
```
Comment (moderakh) [inline commons/src/main/java/com/microsoft/azure/cosmosdb/internal/HttpConstants.java:328]: we should use a substatus code which doesn't collide with the service side substatus codes otherwise the client-retry-policy will be confused.
